src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java

```diff
@@ -0,0 +1,407 @@
+    }
+
+    public static String getApiVersion(String apiVersion) { 
+        return API_VERSION_PREFIX + apiVersion; 
+    }
+
+    private String getPullRequestUrl(String azureUrl, String azureProjectId, String azureRepositoryName, String pullRequestId) {
+        // GET https://{instance}/{collection}/{project}/_apis/git/repositories/{repositoryId}/pullRequests/{pullRequestId}
+        return azureUrl + azureProjectId +
+                "/_apis/git/repositories/" +
+                azureRepositoryName +
+                "/pullRequests/" +
+                pullRequestId;
+    }
+
+    private String getStatusApiUrl(String azureUrl, String azureProjectId, String azureRepositoryName, String pullRequestId, String apiVersion) {
+        // POST https://{instance}/{project}/_apis/git/repositories/{repositoryId}/pullRequests/{pullRequestId}/statuses?api-version=6.0-preview.1
+        return azureUrl + azureProjectId +
+                "/_apis/git/repositories/" +
+                azureRepositoryName +
+                "/pullRequests/" +
+                pullRequestId +
+                "/statuses" +
+                getApiVersion(apiVersion);
+    }
+
+    private String getThreadApiUrl(String azureUrl, String azureProjectId, String azureRepositoryName, String pullRequestId, String apiVersion) {
+        //POST https://{instance}/{project}/_apis/git/repositories/{repositoryId}/pullRequests/{pullRequestId}/threads?api-version=6.0-preview.1
+        return azureUrl + azureProjectId +
+                "/_apis/git/repositories/" +
+                azureRepositoryName +
+                "/pullRequests/" +
+                pullRequestId +
+                "/threads" +
+                getApiVersion(apiVersion);
+    }
+
+    private String getGitPullRequestStatus(AnalysisDetails analysisDetails) throws IOException {
+        final String gitStatusContextGenre = "SonarQube";
+        final String gitStatusContextName = "QualityGate";
+        final String gitStatusDescription = "SonarQube Gate";
+
+        GitPullRequestStatus status = new GitPullRequestStatus(
+                GitStatusStateMapper.toGitStatusState(analysisDetails.getQualityGateStatus()),
+                gitStatusDescription,
+                new GitStatusContext(gitStatusContextGenre, gitStatusContextName),
+                String.format("%s/dashboard?id=%s&pullRequest=%s", server.getPublicRootUrl(),
+                        URLEncoder.encode(analysisDetails.getAnalysisProjectKey(),
+                                StandardCharsets.UTF_8.name()),
+                        URLEncoder.encode(analysisDetails.getBranchName(),
+                                StandardCharsets.UTF_8.name())
+                )
+        );
+        return new ObjectMapper().writeValueAsString(status);
+    }
+
+    private void sendPost(String apiUrl, String body, String successMessage, String authorizationHeader) throws IOException, MalformedURLException, URISyntaxException {
```

Comment:
`IOException` covers `URISyntaxException` and `MalformedURLException`, so only `IOException` needs to be defined here

Reply:
Changed to remove specific exceptions here